File: src/main/java/edu/opjms/candidateSelector/main/ListData.java
package edu.opjms.candidateSelector.main;

import edu.opjms.candidateSelector.util.HouseIndex;
import javafx.collections.ObservableList;

import java.io.Externalizable;
import java.io.IOException;
import java.io.ObjectInput;
import java.io.ObjectOutput;
import java.util.List;

import static edu.opjms.candidateSelector.controls.ActionButtonNew.DEFAULT_NAME;
import static java.util.Arrays.deepToString;
import static javafx.collections.FXCollections.observableArrayList;

class ListData implements Externalizable {
    public static final long serialVersionUID = 234523452312L;
    private final ObservableList<String>[][] candidateList;

    @SuppressWarnings("unchecked")
    public ListData() {
        candidateList = new ObservableList[6][4];
        for (int i = 0; i < 6; i++) {
            for (int j = 0; j < 4; j++) {
                candidateList[i][j] = observableArrayList();
            }
        }
    }

    public ObservableList<String> getCandidateList(HouseIndex houseIndex, int post) {
        return candidateList[houseIndex.ordinal()][post];
    }

    @Override
    public void writeExternal(ObjectOutput out) throws IOException {
        for (int i = 0; i < 6; i++) {
            for (int j = 0; j < 4; j++) {
                candidateList[i][j].removeIf(s -> s.equals(DEFAULT_NAME));
                out.writeObject(List.of(candidateList[i][j].sorted().toArray()));
            }
        }
    }

    @Override
    @SuppressWarnings("unchecked")
    public void readExternal(ObjectInput in) throws IOException, ClassNotFoundException {
        for (int i = 0; i < 6; i++) {
            for (int j = 0; j < 4; j++) {
                candidateList[i][j].setAll((List<String>) in.readObject());
//                System.out.println( in.readObject());
            }
        }
    }

    @Override
    public String toString() {
        return "ListData{" +
                "candidateList=" + deepToString(candidateList) +
                '}';
    }
}

File: src/main/java/edu/opjms/candidateSelector/ListData.kt
package edu.opjms.candidateSelector

import net.kilobyte1000.Houses
import edu.opjms.common.Posts
import javafx.collections.FXCollections
import javafx.collections.ObservableList
import java.io.DataInputStream
import java.io.DataOutputStream
import java.io.IOException
import java.nio.file.Path
import kotlin.io.path.*

@OptIn(ExperimentalPathApi::class)
internal class ListData(path: Path? = null) {

    private val candidateList = if (path != null) {

        val input = DataInputStream(path.inputStream().buffered())

        input.use {
            //used for fast verification, not used
            it.readBoolean()
            List(Houses.SIZE) { _ ->
                val houseMaleList = readList(it)
                val houseFemaleList = readList(it)
                val sportsMaleList = readList(it)
                val sportsFemaleList = readList(it)
                val housePosts = GenderedPost(houseMaleList, houseFemaleList)
                val sportsPost = GenderedPost(sportsMaleList, sportsFemaleList)
                PostPair(housePosts, sportsPost)
            }
        }

    } else {
        List(Houses.SIZE) { PostPair() }
    }

    fun getList(houses: Houses, post: Posts): GenderedPost {
        return if (post == Posts.HOUSE_PREFECT)
            candidateList[houses.ordinal].housePrefects
        else
            candidateList[houses.ordinal].sportsPrefects
    }


    @OptIn(ExperimentalPathApi::class)
    fun writeToFile(path: Path) {
        if (path.notExists()) {
            try {
                path.createDirectories()
            } catch (ignore: IOException) {
            } // file exists
        }

        val out = path.outputStream().buffered()

        var isEmpty = false
        for (posts in candidateList) {
            if (posts.housePrefects.maleList.isEmpty() || posts.housePrefects.femaleList.isEmpty()
                || posts.sportsPrefects.maleList.isEmpty() || posts.sportsPrefects.femaleList.isEmpty()
            ) {
                isEmpty = true
                break
            }
        }
        //used for fast verification if file is OK for use
        DataOutputStream(out).use {
            it.writeBoolean(isEmpty)
            for (pair in candidateList) {
                val (housePrefects, sportPrefects) = pair
                listOf(
                    housePrefects.maleList,
                    housePrefects.femaleList,
                    sportPrefects.maleList,
                    sportPrefects.femaleList
                ).forEach { list ->
                    it.writeInt(list.size)
                    for (item in list) {
                        it.writeUTF(item)
                    }
                }
            }
        }

    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as ListData

        if (candidateList != other.candidateList) return false

        return true
    }

    override fun hashCode(): Int {
        return candidateList.hashCode()
    }


    private fun readList(input: DataInputStream): List<String> {

        val size = input.readInt()
        val list = ArrayList<String>(size)

        repeat(size) {
            val item = input.readUTF()
            list.add(item)
        }
        list.sort()

        // comparision with adjacent strings
        // is enough to check for duplicates in a
        // sorted list
        for (i in size - 1 downTo 1) {
            val name = list[i]
            if (name.isBlank() || name.equals(list[i - 1], true))
                list.removeAt(i)
        }
        return list
    }
}

data class PostPair(
    val housePrefects: GenderedPost = GenderedPost(),
    val sportsPrefects: GenderedPost = GenderedPost()
)

class GenderedPost(maleList: List<String> = mutableListOf(), femaleList: List<String> = mutableListOf()) {
    val maleList: ObservableList<String> = FXCollections.observableArrayList(maleList)
    val femaleList: ObservableList<String> = FXCollections.observableArrayList(femaleList)
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as GenderedPost

        if (maleList != other.maleList) return false
        if (femaleList != other.femaleList) return false

        return true
    }

    override fun hashCode(): Int {
        var result = maleList.hashCode()
        result = 31 * result + femaleList.hashCode()
        return result
    }

    override fun toString(): String {
        return "GenderedPost(maleList=$maleList, femaleList=$femaleList)"
    }


}